File: Hard/2025-12-27-2402-Meeting-Rooms-III/solution.java
import java.util.*;

public class Solution {
    public int mostBooked(int n, int[][] meetings) {
        Arrays.sort(meetings, Comparator.comparingInt(a -> a[0]));
        PriorityQueue<int[]> freeRooms = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        PriorityQueue<int[]> busyRooms = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        int[] roomCount = new int[n];
        
        for (int i = 0; i < n; i++) {
            freeRooms.offer(new int[]{0, i}); // {end time, room number}
        }
        
        for (int[] meeting : meetings) {
            int start = meeting[0];
            int end = meeting[1];
            
            while (!busyRooms.isEmpty() && busyRooms.peek()[0] <= start) {
                int[] room = busyRooms.poll();
                freeRooms.offer(room);
            }
            
            if (!freeRooms.isEmpty()) {
                int[] room = freeRooms.poll();
                roomCount[room[1]]++;
                busyRooms.offer(new int[]{end, room[1]});
            } else {
                int[] room = busyRooms.poll();
                int newStart = room[0];
                roomCount[room[1]]++;
                busyRooms.offer(new int[]{newStart + (end - start), room[1]});
            }
        }
        
        int maxMeetings = 0;
        int roomWithMaxMeetings = 0;
        
        for (int i = 0; i < n; i++) {
            if (roomCount[i] > maxMeetings) {
                maxMeetings = roomCount[i];
                roomWithMaxMeetings = i;
            }
        }
        
        return roomWithMaxMeetings;
    }
}

File: Hard/2025-12-27-2402-Meeting-Rooms-III/solution.py
import heapq
from collections import defaultdict

class Solution:
    def mostBooked(self, n: int, meetings: List[List[int]]) -> int:
        meetings.sort()
        room_count = [0] * n
        free_rooms = list(range(n))
        busy_rooms = []
        
        for start, end in meetings:
            while busy_rooms and busy_rooms[0][0] <= start:
                finish_time, room = heapq.heappop(busy_rooms)
                heapq.heappush(free_rooms, room)
            
            if free_rooms:
                room = heapq.heappop(free_rooms)
                room_count[room] += 1
                heapq.heappush(busy_rooms, (end, room))
            else:
                finish_time, room = heapq.heappop(busy_rooms)
                room_count[room] += 1
                delay = finish_time - start
                new_end = finish_time + (end - start)
                heapq.heappush(busy_rooms, (new_end, room))
        
        return room_count.index(max(room_count))